### PR TITLE
Membership end_date is not correctly set with 2 or more initial membership terms

### PR DESCRIPTION
**Pull request by agilewarealok**
Overview
----------------------------------------
When we create a Membership with 2 or more initial membership terms. The initial membership end date is set correctly in CiviCRM.

However, if the related Contribution has a status of Pending, when this Contribution status is changed to Completed then the membership end date is re-calculated and using the default term, causing the end date to change.

Put simply, when user sign up with pay later option for a 3 year membership. When the payment is received this membership is then changed to 1 year when Contribution is marked as Completed.

Before - Steps to Reproduce
----------------------------------------
1. Open a Contact Page
2. Go to Membership > Create a Membership with 2 or more terms with Contribution status of Pending.
3. Verify above steps created one pending membership and one pending contribution.
4. Go to Contributions tab > Click Record Payment of latest pending contribution
5. Record all the due payment
6. Go to Memberships tab
7. Membership is active but end_date is not correct. end_date is calculated for 1 term only.

After
----------------------------------------
On Step 7 Membership end_date is set according to the selected number of terms.

Comments
----------------------------------------
_Agileware Ref: CIVICRM-1148_

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@agilewarealok  - jenkins is posting a couple of style warnings

**Comment by mattwire** [comment]
@agilewarealok Can we trade reviews on this? https://github.com/civicrm/civicrm-core/pull/12542 has been around for a while now..

**Comment by agilewarealok** [comment]
jenkins re-test this please

**Comment by agilewarealok** [comment]
jenkins re-test this please

**Comment by agilewarealok** [comment]
@mattwire Sure, sounds good.

**Comment by jusfreeman** [comment]
@eileenmcnaughton would you mind providing some guidance as to what needs to happen here to progress this PR? Seems it is stuck again.

**Comment by eileenmcnaughton** [comment]
It has a test so it's in the top 30 priorities for review - I don't know how to generate more reviewer capacity though - that's the bottleneck.

**Comment by JoeMurray** [comment]
We typically ask someone to trade a review. Unfortunately we're way overloaded right now.

File: CRM/Contribute/BAO/Contribution.php [hunk 1]
```diff
@@ -467,14 +467,29 @@ public static function getPaymentProcessorReadyAddressParams($params, $billingLo
...
+    $membershipTerms = 1;
+
+    while ($termsDAO->fetch()) {
+      $qty = $termsDAO->qty;
+      $numTerms = $termsDAO->membership_num_terms;
+
+      if (!empty($qty)) {
+        $membershipTerms = $qty;
+      }
+      elseif (empty($numTerms)) {
```
Comment (seamuslee001) [inline CRM/Contribute/BAO/Contribution.php]: should this be !empty? @jusfreeman

**Comment by eileenmcnaughton** [comment]
I took a quick look and found that when paying by credit card (dummy processor) the date is miscalculated without this patch - 2*1 year term = 4 and with this patch 2*1 year term = 1!!

I'll dig a bit more later

File: CRM/Contribute/BAO/Contribution.php (hunk 1 above)
Comment (agilewarealok) [reply CRM/Contribute/BAO/Contribution.php]: @seamuslee001 Yes it should be **!empty**. Fixed and pushed the changes.

**Comment by eileenmcnaughton** [comment]
@agilewarealok in general my thinking on this is that we have 2 scenarios

1) new membership
2) renewal

In the case of the former the end date should be set correctly when it is created as 'new' & so on complete it will be a status change not a date change. In the latter case the date will be calculated on complete.

In general I think this means it's a pretty fragile bit of logic & we'll have to look at what scenarios exist and what test cover there is when making a change. 

One thing we have been working on that should simplify it slightly is getting rid of the separate logic in the AdditionalPayment form & using the payment api there - so there are less code paths in play

**Comment by agilewarealok** [comment]
@eileenmcnaughton What are the next steps on this to proceed further ?

**Comment by eileenmcnaughton** [comment]
@agilewarealok I'm not sure - I need to find a good chunk of time to really dig into it

File: CRM/Price/BAO/LineItem.php [hunk 2]
```diff
@@ -371,15 +372,15 @@ public static function format($fid, $params, $fields, &$values, $amount_override
...
+        'line_total' => $lineTotal,
         'participant_count' => $qty * $participantsPerField,
         'max_value' => CRM_Utils_Array::value('max_value', $options[$oid]),
         'membership_type_id' => CRM_Utils_Array::value('membership_type_id', $options[$oid]),
         'membership_num_terms' => CRM_Utils_Array::value('membership_num_terms', $options[$oid]),
         'auto_renew' => CRM_Utils_Array::value('auto_renew', $options[$oid]),
         'html_type' => $fields['html_type'],
         'financial_type_id' => CRM_Utils_Array::value('financial_type_id', $options[$oid]),
-        'tax_amount' => CRM_Utils_Array::value('tax_amount', $options[$oid]),
+        'tax_amount' => (CRM_Utils_Array::value('tax_amount', $options[$oid])),
```
Comment (mlutfy) [inline CRM/Price/BAO/LineItem.php]: Are these parenthesis necessary?

**Comment by jusfreeman** [comment]
@seamuslee001 @eileenmcnaughton this PR is relatively straight forward, fixes a reproducible bug and has a unit test.

We will update the PR based on feedback received. Once that's done would you mind reviewing again with a view to merging?

We've been using this patch in our CiviCRM build since February without any related issues and the bug is fixed.

**Comment by eileenmcnaughton** [comment]
@jusfreeman when I tested it - [back in March](https://github.com/civicrm/civicrm-core/pull/13706#issuecomment-478165244) - I found that it fixed one bug & created another 

So I think we need to make sure we've got a full test grid of scenarios we are testing against

**Comment by eileenmcnaughton** [comment]
This is stale. SInce it has a test I rescued the test & put up a more conservative alternative
https://github.com/civicrm/civicrm-core/pull/14902

**Comment by jusfreeman** [comment]
@eileenmcnaughton any chance you can post review comments on a PR instead of calling it stale and creating a new one? We would have merged any changes required.

Just seems weird to close a PR only to open a new one with the minor changes and then merge it. Whereas most of the discussion and the majority of work was actually expended here, on this PR.

The vibe does not seem right.

**Comment by eileenmcnaughton** [comment]
@jusfreeman generally when I come across a PR that is stale that has a unit test I try to rescue the unit test & get that merged. If it doesn't have a test I close it & let the person re-open it when it is not stale anymore

People are always welcome to rebase their PR & re-open it - closing it in the meantime helps to get us to a situation where the PRs in the PR queue are reviewable - generally when I go to do review I spend more time finding PRs that are actually in a reviewable state 

We have discussed a few times the idea of abandoning trying to actively review the PR queue (which would really mean people just review things that scratch their own itch/ where they can trade review) - if we did that it would mean it wouldn't matter if the review queue is full of things that are not review ready (but also we would be reviewing a smaller % of the ones that come in). I go back & forth on this.

**Pull request closed without merging**